### ISSUE TITLE
Value deserialization error

### ISSUE THREAD

**Issue by vishwajeets7** [ch-robinson/dotnet-avro#118]
I have a DB2 source connector publishing data to a topic, the schema of the topic contains fields like...

"fields":[{"name":"ID", "type":"int","logicalType":"INTEGER", "dbColumnName":"ID"},"default":0},
.....

with the chr.avro.confluent versions 7.0.3 and 7.0.4 throws a deserialization error, unable to proceed further. (all the dependencies are added to the project)

Error:
Inner Exception 1:
SchemaRegistryException: Schema 304 not found io.confluent.rest.exceptions.RestNotFoundException: Schema 304 not found io.confluent.rest.exceptions.RestNotFoundException: Schema 304 not found
...
...
error code: 40403

please note, i'm running the consumer for the topic created by the source connector.

Thanks,
Vishu

**Comment by dstelljes** [comment]
This is probably an issue with the connector, not Chr.Avro. Can you provide a redacted connector configuration?

**Comment by vishwajeets7** [comment]
@dstelljes 
Thanks for your response.

This is IIDR (IBM) to Kafka, i have no access to the configuration however i can get an answer in case you have specific question.

My client is .net core 3.1, let me add to the issue.

**Comment by dstelljes** [comment]
Specifically, it'd be helpful to know the connector class and the values of the `key.converter` and `value.converter` properties. You could also validate that the Schema Registry URLs are set properly. It seems likely that the connector isn't serializing data according to the Confluent wire format.

**Comment by vishwajeets7** [comment]
@dstelljes 
I think the converter value is Avro, but will confirm it.

Related to Logical Type being included in the schema definition, have seen this in other use cases as well. Confluent kafka github has example where the schema definition includes the Logical Type and it works as expected.

https://github.com/confluentinc/confluent-kafka-dotnet/blob/master/examples/AvroSpecific/User.cs

**Comment by dstelljes** [comment]
If it were a logical type problem, you wouldn’t be seeing the `Schema 304 not found` exception. One other question, are you expecting both keys and values to be Avro?

**Comment by vishwajeets7** [comment]
key.converter & value.converter are both set to "io.confluent.connect.avro.AvroConverter"

**Comment by dstelljes** [comment]
You may want to remove the `schemas.enable` properties; `schemas.enable` instructs Connect to include the schema with each message: https://www.confluent.io/blog/kafka-connect-deep-dive-converters-serialization-explained/#json-schemas

**Comment by vishwajeets7** [comment]
Appreciate your response @dstelljes 
I will check & post the outcome soon. 
Thanks.

**Comment by vishwajeets7** [comment]
Meanwhile the schema configuration check is being worked on, a quick question the same schema is able to read with Java client successfully. Curious to know as the schema registry is url is same used by both the consumers, however the .net consumer is facing deserialization error while Java consumer could deserialize successfully.

**Comment by dstelljes** [comment]
Can you post the code you’re using to create the consumer and deserializers in the .NET client?

**Comment by vishwajeets7** [comment]
code is on different machine, so manually typing...please ignore the syntax & spelling errors.

Var consumerConfig = new ConsumerConfig
{
BootstrapServers = xxx,
GroupId = xxx,
SecuirityProtocol = SSL,
SslCaLocation = xxx,
SslCertificateLocation = xxx,
SslKeyLocation = xxx,
AutoOffsetReset = Earilest,
EnableAutoCommit= true,
EnablePartitionEof = true
}

using(var registry = CachedSchemaRegistry(regConfig))
{
var builder = new ConsumerBuilder<Ignore, MODEL>(consumerConfig)
.setAvroValueDeserializer(registry)

using(var consumer = builder.build())
{
consumer.subscribe(topicname);

while(true)
{
 var res = consumer.consume();
....
}
}
}

**Comment by dstelljes** [comment]
Thanks, that all looks good. Is the Java consumer configured with the same `auto.offset.reset`? You may want to validate that the Java consumer is also able to read all records from the beginning of the topic; if not, that would help narrow down a cause.

**Comment by vishwajeets7** [comment]
auto.offset.reset is set to Earliest for java client as well and could read the message from beginning.

**Comment by dstelljes** [comment]
Interesting. I’m at a loss as to why `SchemaRegistryException: Schema 304 not found` would be thrown if the Java client is pointed at the same registry.

**Comment by vishwajeets7** [comment]
That confuses me too. Is there anything on the .net consumer that can be configured and i am missing? how deserialization can be handled?

Have tested the similar code with .net producer & consumer, which works great.

**Comment by dstelljes** [comment]
I don’t think so; the exception you’re seeing is being thrown here before the Avro deserialization begins: https://github.com/ch-robinson/dotnet-avro/blob/7.0.4/src/Chr.Avro.Confluent/AsyncSchemaRegistryDeserializer.cs#L174

**Comment by vishwajeets7** [comment]
I am pretty sure it is related to deserialization error after the schema is pulled, as when the deserializer is not used i could receive the data with unicode characters appended to the message values with ConsumerBuilder<string, string>

Example : \u0004PM \u0002N

**Comment by dstelljes** [comment]
Avro prepends strings with their length, so the \u0004 and \u0002 you're seeing there are the lengths (varint-encoded) of the strings they precede. The Schema Registry deserializer expects that every payload follows the Confluent wire format, which expects the first 5 bytes of the payload to be 0x00 followed by the schema ID.

Could you provide the code that sets up your working Java consumer?

**Comment by vishwajeets7** [comment]
The unicode values pasted here are just an example, i do know the magic
bytes and could see them appended to the first column value.

Sure i will try to post the java client code as well.

Thanks

**Comment by vishwajeets7** [comment]
didn't get much details about the code, Java client is using GenericAvroSerde, if it helps.
